app/src/main/java/com/keylesspalace/tusky/EditProfileActivity.kt

```diff
@@ -94,93 +88,133 @@ class EditProfileActivity : BaseActivity(), Injectable {
+
+                        accountFieldEditAdapter.setFields(me.source?.fields ?: emptyList())
+
+                        if(viewModel.avatarData.value == null) {
+                            Picasso.with(this@EditProfileActivity)
+                                    .load(me.avatar)
+                                    .placeholder(R.drawable.avatar_default)
+                                    .into(avatarPreview)
+                        }
+
+                        if(viewModel.headerData.value == null) {
+                            Picasso.with(this@EditProfileActivity)
+                                    .load(me.header)
+                                    .into(headerPreview)
+                        }
+
+                    }
                 }
-                if (!headerChanged) {
-                    Picasso.with(headerPreview.context)
-                            .load(me.header)
-                            .into(headerPreview)
+                is Error -> {
+                    val snackbar = Snackbar.make(avatarButton, R.string.error_generic, Snackbar.LENGTH_LONG);
+                    snackbar.setAction(R.string.action_retry) {
+                        viewModel.obtainProfile()
+                    }
+                    snackbar.show()
+
                 }
             }
+        })
+
+        observeImage(viewModel.avatarData, avatarPreview, avatarProgressBar)
+        observeImage(viewModel.headerData, headerPreview, headerProgressBar)
 
-            override fun onFailure(call: Call<Account>, t: Throwable) {
-                onAccountVerifyCredentialsFailed()
+        viewModel.saveData.observe(this, Observer<Resource<Boolean>> {
+            when(it) {
+                is Success -> {
+                    if(it.data == true) {
+                        privatePreferences.edit()
+                                .putBoolean("refreshProfileHeader", true)
+                                .apply()
+                    }
+                    setResult(Activity.RESULT_OK)
+                    finish()
+                }
+                is Loading -> {
+                    saveProgressBar.visibility = View.VISIBLE
+                }
+                is Error -> {
+                    onSaveFailure(it.errorMessage)
+                }
             }
         })
+
     }
 
-    override fun onSaveInstanceState(outState: Bundle) {
-        outState.run {
-            putString(KEY_OLD_DISPLAY_NAME, oldDisplayName)
-            putString(KEY_OLD_NOTE, oldNote)
-            putBoolean(KEY_OLD_LOCKED, oldLocked)
-            putBoolean(KEY_IS_SAVING, isSaving)
-            putSerializable(KEY_CURRENTLY_PICKING, currentlyPicking)
-            putBoolean(KEY_AVATAR_CHANGED, avatarChanged)
-            putBoolean(KEY_HEADER_CHANGED, headerChanged)
+    override fun onStop() {
+        super.onStop()
+        if(!isFinishing) {
+            viewModel.updateProfile(displayNameEditText.text.toString(),
```

Comment:
Why not update immediately after change?

Reply:
because that would require onTextChanged listeners on all input fields